client/beefy/src/worker.rs

```diff
@@ -351,55 +349,171 @@ where
-					)
-					.is_err()
-				{
-					// just a trace, because until the round lifecycle is improved, we will
-					// conclude certain rounds multiple times.
-					trace!(target: "beefy", "🥩 Failed to append justification: {:?}", signed_commitment);
+				if let Err(e) = self.backend.append_justification(
+					BlockId::Number(block_num),
+					(
+						BEEFY_ENGINE_ID,
+						VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+					),
+				) {
+					trace!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, signed_commitment);
 				}
 				self.signed_commitment_sender
 					.notify(|| Ok::<_, ()>(signed_commitment))
 					.expect("forwards closure result; the closure always returns Ok; qed.");
 
-				self.best_beefy_block = Some(block_num);
-				if let Err(err) = self.client.hash(block_num).map(|h| {
-					if let Some(hash) = h {
-						self.beefy_best_block_sender
-							.notify(|| Ok::<_, ()>(hash))
-							.expect("forwards closure result; the closure always returns Ok; qed.");
-					}
-				}) {
-					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
-						block_num, err);
+				self.set_best_beefy_block(block_num);
+
+				// Vote if there's now a new vote target.
+				if let Some(target_number) = self.current_vote_target() {
+					self.do_vote(target_number);
 				}
+			}
+		}
+	}
+
+	/// Create and gossip Signed Commitment for block number `target_number`.
+	///
+	/// Also handle this self vote by calling `self.handle_vote()` for it.
+	fn do_vote(&mut self, target_number: NumberFor<B>) {
+		trace!(target: "beefy", "🥩 Try voting on {}", target_number);
+
+		// Most of the time we get here, `target` is actually `best_grandpa`,
+		// avoid asking `client` for header in that case.
+		let target_header = if target_number == *self.best_grandpa_block_header.number() {
+			self.best_grandpa_block_header.clone()
+		} else {
+			match self.client.expect_header(BlockId::Number(target_number)) {
+				Ok(h) => h,
+				Err(err) => {
+					debug!(
+						target: "beefy",
+						"🥩 Could not get header for block #{:?} (error: {:?}), skipping vote..",
+						target_number,
+						err
+					);
+					return
+				},
+			}
+		};
+		let target_hash = target_header.hash();
+
+		let mmr_root = if let Some(hash) = self.extract_mmr_root_digest(&target_header) {
+			hash
+		} else {
+			warn!(target: "beefy", "🥩 No MMR root digest found for: {:?}", target_hash);
+			return
+		};
+		let payload = Payload::new(known_payload_ids::MMR_ROOT_ID, mmr_root.encode());
+
+		let (validators, validator_set_id) = if let Some(rounds) = &self.rounds {
+			if !rounds.should_self_vote(&(payload.clone(), target_number)) {
+				debug!(target: "beefy", "🥩 Don't double vote for block number: {:?}", target_number);
+				return
+			}
+			(rounds.validators_for(target_number), rounds.validator_set_id_for(target_number))
+		} else {
+			debug!(target: "beefy", "🥩 Missing validator set - can't vote for: {:?}", target_hash);
+			return
+		};
```

Comment:
The fact that this implicitly votes with different validator sets is a bit confusing IMO, it also ties with the logic of `vote_target`. But let's keep it as is for now and refactor in the future once we block until we vote for mandatory blocks.